src/DwhManager.php

```diff
@@ -0,0 +1,414 @@
+    {
+        $this->connection->grantOnDatabaseToRole(
+            $this->database,
+            $role,
+            $databasePrivileges
+        );
+        $this->logger->info(sprintf(
+            'Granted [%s] to role "%s" on database "%s"',
+            implode(',', $databasePrivileges),
+            $role,
+            $this->database
+        ));
+    }
+
+    private function ensureRoleHasSchemaPrivileges(string $role, array $schemaPrivileges, string $schemaName): void
+    {
+        $this->connection->grantOnSchemaToRole($schemaName, $role, $schemaPrivileges);
+        $this->logger->info(sprintf(
+            'Granted [%s] to role "%s" on schema "%s"',
+            implode(',', $schemaPrivileges),
+            $role,
+            $schemaName
+        ));
+    }
+
+    private function ensureRoleHasWarehousePrivileges(string $role, array $warehousePrivileges): void
+    {
+        $this->connection->grantOnWarehouseToRole(
+            $this->warehouse,
+            $role,
+            $warehousePrivileges
+        );
+        $this->logger->info(sprintf(
+            'Granted [%s] to role "%s" on warehouse "%s"',
+            implode(',', $warehousePrivileges),
+            $role,
+            $this->warehouse
+        ));
+    }
+
+    private function ensureRoleOnlyHasRolesGranted(string $roleName, array $desiredRoles): void
+    {
+        $grantedRoles = $this->checker->getGrantedRolesOfRole($roleName);
+        $toBeRevoked = array_diff($grantedRoles, $desiredRoles);
+        $toBeGranted = array_diff($desiredRoles, $grantedRoles);
+        foreach ($toBeRevoked as $revokedRole) {
+            $this->connection->revokeRoleGrantFromRole($revokedRole, $roleName);
+            $this->logger->info(sprintf(
+                'Role "%s" has been revoked from role "%s"',
+                $revokedRole,
+                $roleName
+            ));
+        }
+        foreach ($toBeGranted as $grantedRole) {
+            $this->connection->grantRoleToRole($grantedRole, $roleName);
+            $this->logger->info(sprintf(
+                'Role "%s" has been granted to role "%s"',
+                $grantedRole,
+                $roleName
+            ));
+        }
+    }
+
+    private function ensureSchemaExists(string $schemaName): void
+    {
+        if (!$this->checker->existsSchema($schemaName)) {
+            $this->connection->createSchema($schemaName);
+            $this->logger->info(sprintf(
+                'Created schema "%s"',
+                $schemaName
+            ));
+        } else {
+            $this->logger->info(sprintf(
+                'Schema "%s" exists',
+                $schemaName
+            ));
+        }
+    }
+
+    private function ensureUserExists(string $userName, array $options): void
+    {
+        if (!$this->checker->existsUser($userName)) {
+            $password = $this->generatePassword();
+            $this->connection->createUser(
```

Comment:
Nebudeme nastavovat všem uživatelům `MUST_CHANGE_PASSWORD = true`? https://docs.snowflake.net/manuals/sql-reference/sql/create-user.html
To heslo zůstane v tom logu a zároveň si ho budou muset nějak přepad tak by imho bylo fajn jakmile ho dostane cílový uživatel aby si ho musel změnit.